server/services/v1/database/query_index.go

```diff
@@ -0,0 +1,418 @@
+// Copyright 2022-2023 Tigris Data, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package database
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/buger/jsonparser"
+	"github.com/rs/zerolog/log"
+	api "github.com/tigrisdata/tigris/api/server/v1"
+	"github.com/tigrisdata/tigris/errors"
+	"github.com/tigrisdata/tigris/internal"
+	"github.com/tigrisdata/tigris/keys"
+	"github.com/tigrisdata/tigris/query/filter"
+	"github.com/tigrisdata/tigris/schema"
+	"github.com/tigrisdata/tigris/server/transaction"
+	"github.com/tigrisdata/tigris/store/kv"
+	"github.com/tigrisdata/tigris/value"
+)
+
+type QueryIndexReader struct {
+	ctx      context.Context
+	coll     *schema.DefaultCollection
+	filter   *filter.WrappedFilter
+	tx       transaction.Tx
+	err      error
+	keyRange []keys.Key
+	kvIter   kv.Iterator
+}
+
+func NewQueryIndexReader(ctx context.Context, tx transaction.Tx, coll *schema.DefaultCollection, filter *filter.WrappedFilter, keyRange []keys.Key) *QueryIndexReader {
+	return &QueryIndexReader{
+		ctx:      ctx,
+		tx:       tx,
+		coll:     coll,
+		filter:   filter,
+		err:      nil,
+		keyRange: keyRange,
+	}
+}
+func (reader *QueryIndexReader) CreateIter() (*QueryIndexReader, error) {
+	var err error
+	switch len(reader.keyRange) {
+	case 0:
+		return nil, errors.InvalidArgument("Index Key range cannot be zero, we need a minimum of one key")
+	case 1:
+		reader.kvIter, err = reader.tx.Read(reader.ctx, reader.keyRange[0])
+		if err != nil {
+			return nil, err
+		}
+	case 2:
+		reader.kvIter, err = reader.tx.ReadRange(reader.ctx, reader.keyRange[0], reader.keyRange[1], false)
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, errors.InvalidArgument("Incorrectly created query key range")
+	}
+
+	return reader, nil
+}
+
+func BuildQueryIndexKeys(coll *schema.DefaultCollection, queryFilters []filter.Filter) ([]keys.Key, error) {
```

Comment:
I am thinking if we move the entire key building logic in `key_builder.go`. As we have the logic of building primary key rows in key builder then better to have building secondary index keys logic at the same place.